Aggregate PRS Output

michaelgatzen opened the pull request:
- PRSWrapper.wdl
  - Added columns `batch_id` and `is_control` to results.csv
- New file: AggregatePRSResults
  - Aggregate multiple results CSV files into summarized tables
    - Check that all `batch_id`s are the same
    - Check that there is exactly one control sample within every batch
    - Extract control sample
    - Extract results in two shapes: `all_results` and `summarized_results`
    - Extract score distribution
  - Create population PCA plot
  - Build HTML report
    - Compare control sample to expected control result
    - Batch summary
    - Score distribution
    - Population PCA
    - Individual results
- Added to Dockstore.yml
- Fixed bug in ScoringPart.wdl

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -0,0 +1,235 @@
+
+    num_control_samples <- sum(results$is_control=="true")
+
+    if (num_batches != 1) {
```
`num_batches` should be `num_controls`

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -0,0 +1,235 @@
+
+    write_tsv(results, paste0(batch_id, "_all_results.tsv"))
+
+    write_tsv(results %>% filter(is_control == "true"), paste0(batch_id, "_control_results.tsv"))
```
should be able to just do `filter(is_control)`.  "true" should be parsed as a logical by read_tsv.

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -0,0 +1,235 @@
+
+    write_tsv(results %>% filter(is_control == "true"), paste0(batch_id, "_control_results.tsv"))
+
+    results_pivoted <- results %>% select(-batch_id) %>% select(-is_control) %>% pivot_longer(!sample_id, names_to=c("condition",".value"), names_pattern="([^_]+)_(.+)")
```
use a single select `select(-batch_id, -is_control)`

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -0,0 +1,235 @@
+
+    ## Individual Sample Results
+    \`\`\`{r sample results , echo = FALSE, results = "asis"}
+    kable(batch_all_results %>% mutate(across(ends_with("risk"), ~ kableExtra::cell_spec(.x, color=ifelse(is.na(.x), "blue", ifelse(.x=="NOT_RESULTED", "red", ifelse(.x == "HIGH", "orange", "green")))))), digits = 2)
```
I think we should filter out the control sample from batch_all_results before displaying the table

kachulis reviewed (changes requested):
Looks great @michaelgatzen!  some minor comments about using more dplyr.

michaelgatzen commented:
Addressed all comments and made a few changes to the HTML report:
- Added column to the control table describing which one is the batch and which one is the control sample
- Removed `is_control` column from the Individual Sample Results table

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -198,7 +200,7 @@ task BuildHTMLReport {
     ## Control Sample
     \`\`\`{r control, echo = FALSE, results = "asis"}
-    kable(list(batch_control_results, expected_control_results) %>% reduce(bind_rows) %>% select(-batch_id) %>% select(-is_control) , digits = 2)
+    kable(list(batch_control_results, expected_control_results) %>% reduce(bind_rows) %>% select(-batch_id, -is_control) %>% add_column(Sample=c('Batch control', 'Expected control'), .before='sample_id'), digits = 2)
```
I think we should drop the sample_id column as well here.  All we care about is which is the batch control, and which is the expected control, the sample id of the control is unimportant.

kachulis commented on ImputationPipeline/AggregatePRSResults.wdl:
```diff
@@ -57,27 +57,27 @@ task AggregateResults {
     }
 
-    batch_id <- results$batch_id[1]
+    batch_id <- batch_ids[1]
```
you don't actually need to select the particular element here.  In R a single value is actually vector of length 1.  So, since batch_ids is guaranteed to be length 1 a this point in the code, `batch_ids`, `batch_ids[1]`, and `batch_ids[[1]]` are all the same.

michaelgatzen replied:
Wizardry!

kachulis reviewed (changes requested):
thanks @michaelgatzen , just two very small additional requests.
